propolis-client/src/lib.rs

```diff
@@ -1,6 +1,152 @@
 //! Interface for making API requests to propolis.
 //! This should be replaced with a client generated from the OpenAPI spec.
 
+use reqwest::Body;
+use serde::de::DeserializeOwned;
+use slog::{info, o, Logger};
+use std::net::SocketAddr;
+use thiserror::Error;
+use uuid::Uuid;
+
 pub mod api;
 
-// TODO: Implement the client.
+/// Errors which may be returend from the Propolis Client.
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("Request failed: {0}")]
+    Reqwest(#[from] reqwest::Error),
+
+    #[error("Bad Status: {0}")]
+    Status(u16),
+}
+
+/// Client-side connection to propolis.
+pub struct Client {
+    client: reqwest::Client,
+    log: Logger,
+    address: SocketAddr,
+}
+
+// Sends "request", awaits "response", and returns an error on any
+// non-success status code.
+//
+// TODO: Do we want to handle re-directs?
+async fn send_and_check_ok(
+    request: reqwest::RequestBuilder,
+) -> Result<reqwest::Response, Error> {
+    let response = request.send().await.map_err(Error::from)?;
+
+    if !response.status().is_success() {
+        return Err(Error::Status(response.status().as_u16()));
+    }
+
+    Ok(response)
+}
+
+// Sends a "request", awaits "response", and parses the body
+// into a deserializable type.
+async fn send_and_parse_response<T: DeserializeOwned>(
+    request: reqwest::RequestBuilder,
+) -> Result<T, Error> {
+    send_and_check_ok(request).await?.json().await.map_err(|e| e.into())
+}
+
+impl Client {
+    pub fn new(address: SocketAddr, log: Logger) -> Client {
```

Comment:
Should this take a `T: ToSocketAddrs`? Makes it a little more ergonomic in what a user can pass in.

Reply:
Hrmmm wouldn't this imply accepting multiple addresses, rather than just one? Seems like this would split the invocations to reqwest into multiple invocations (try one address, if it fails, try the next, etc). Is there a good wrapper for this?

Reply:
That does touch on another point I guess: the reason it returns an iterator is because `ToSocketAddrs` is implemented for strings for which it may end up needing to do name resolution and that might end up returning multiple results. It's a good question in whether we want `Client::new` to possibly be performing name resolution as well. If we don't, then I'm fine with leaving it as is and then the caller can decide to just call `.to_socket_addrs()?.next()` if they want.